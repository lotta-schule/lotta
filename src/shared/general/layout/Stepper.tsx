import * as React from 'react';
import { KeyboardArrowLeft, KeyboardArrowRight } from '@material-ui/icons';
import { NavigationButton } from '../button/NavigationButton';
import clsx from 'clsx';

import styles from './Stepper.module.scss';

export interface StepperProps {
    className?: string;
    style?: React.CSSProperties;
    maxSteps: number;
    currentStep: number;
    onStep: (newStep: number) => void;
}

export const Stepper = React.memo<StepperProps>(
    ({ className, currentStep, maxSteps, style, onStep }) => {
        return (
            <div className={clsx(styles.root, className)} style={style}>
                <NavigationButton
                    small
                    icon={<KeyboardArrowLeft />}
                    onClick={() => onStep(currentStep - 1)}
                    disabled={currentStep <= 0}
                    aria-label={'Vorheriger Schritt'}
                >
                    vorheriges
                </NavigationButton>
                <div>
                    {currentStep + 1} / {maxSteps}
                </div>
                <NavigationButton
                    small
                    icon={<KeyboardArrowRight />}
                    onClick={() => onStep(currentStep + 1)}
                    disabled={currentStep >= maxSteps - 1}
                    aria-label={'Nächster Schritt'}
                >
                    nächstes
                </NavigationButton>
            </div>
        );
    }
);
Stepper.displayName = 'Stepper';
