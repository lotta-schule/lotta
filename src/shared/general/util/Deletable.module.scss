.root {
    display: inline-block;
    position: relative;
    .button {
        position: absolute;
        top: -7.5px;
        right: -7.5px;
        width: 15px;
        height: 15px;
        overflow: hidden;
        z-index: 1;
        transform: scale(0);
    }
    &:hover {
        .button {
            transform: scale(1);
        }
    }
}
