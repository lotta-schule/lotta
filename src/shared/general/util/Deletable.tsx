import * as React from 'react';
import { Close } from '@material-ui/icons';
import { Button } from '../button/Button';
import clsx from 'clsx';

import styles from './Deletable.module.scss';

export interface DeletableProps
    extends Omit<React.HTMLProps<HTMLDivElement>, 'onClick'> {
    className?: string;
    title?: string;
    onDelete?: React.MouseEventHandler<HTMLButtonElement> | null;
}

export const Deletable: React.FC<DeletableProps> = ({
    children,
    onDelete,
    title = 'lÃ¶schen',
    className,
    ...props
}) => {
    return (
        <div className={clsx(styles.root, className)} {...props}>
            {onDelete && (
                <Button
                    small
                    className={styles.button}
                    icon={<Close />}
                    title={title}
                    onClick={onDelete ?? undefined}
                />
            )}
            {children}
        </div>
    );
};
