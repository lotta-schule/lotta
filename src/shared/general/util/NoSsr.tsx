import * as React from 'react';

export const NoSsr: React.FC = ({ children }) => {
    const [hasRendered, setHasRendered] = React.useState(false);
    React.useEffect(() => {
        setHasRendered(true);
    }, []);

    return hasRendered ? (children as React.ReactElement) : null;
};
