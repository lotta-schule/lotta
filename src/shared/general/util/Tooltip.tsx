import * as React from 'react';
import { Tooltip as ReachTooltip } from '@reach/tooltip';
import clsx from 'clsx';

import styles from './Tooltip.module.scss';

export interface TooltipProps extends React.HTMLAttributes<HTMLDivElement> {
    label: React.ReactNode;
}

export const Tooltip = React.forwardRef(
    (
        { children, label, className, ...props }: TooltipProps,
        ref: React.ForwardedRef<HTMLDivElement>
    ) => {
        return (
            <ReachTooltip
                ref={ref}
                label={label}
                className={clsx(styles.root, className)}
                {...props}
            >
                {children}
            </ReachTooltip>
        );
    }
);
Tooltip.displayName = 'Tooltip';
