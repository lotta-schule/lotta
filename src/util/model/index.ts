export * from './Tenant';
export * from './User';
export * from './Widget';