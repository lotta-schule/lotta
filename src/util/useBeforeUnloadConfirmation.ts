import { useRouter } from 'next/router';
import * as React from 'react';

export const useBeforeUnloadConfirmation = (
    enabled: boolean,
    message: string
) => {
    const lastHistoryState = React.useRef(global.history?.state);
    const router = useRouter();
    React.useEffect(() => {
        const storeLastHistoryState = () => {
            lastHistoryState.current = history.state;
        };
        router.events.on('routeChangeComplete', storeLastHistoryState);
        return () => {
            router.events.off('routeChangeComplete', storeLastHistoryState);
        };
    }, [router]);

    React.useEffect(() => {
        let isWarned = false;

        const routeChangeStart = (url: string) => {
            if (router.asPath !== url && enabled && !isWarned) {
                isWarned = true;
                if (confirm(message)) {
                    router.push(url);
                } else {
                    isWarned = false;
                    router.events.emit('routeChangeError');

                    // HACK
                    const state = lastHistoryState.current;
                    if (
                        state != null &&
                        history.state != null &&
                        state.idx !== history.state.idx
                    ) {
                        history.go(state.idx < history.state.idx ? -1 : 1);
                    }

                    throw 'Abort route change. Please ignore this error.';
                }
            }
        };

        const beforeUnload = (e: Event) => {
            if (enabled && !isWarned) {
                const event = e || window.event;
                event.returnValue = true;
                return message;
            }
            return null;
        };

        router.events.on('routeChangeStart', routeChangeStart);
        window.addEventListener('beforeunload', beforeUnload);

        return () => {
            router.events.off('routeChangeStart', routeChangeStart);
            window.removeEventListener('beforeunload', beforeUnload);
        };
    }, [message, router, enabled]);
};
