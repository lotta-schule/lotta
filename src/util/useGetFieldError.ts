import { useCallback } from 'react';
import { ApolloError } from '@apollo/client';

export const useGetFieldError = (error?: ApolloError) => {
    return useCallback(
        (fieldName: string): string | false => {
            if (error) {
                const messages: string[] | undefined =
                    error.graphQLErrors &&
                    (error.graphQLErrors.map(
                        (graphQLError: any) =>
                            graphQLError &&
                            graphQLError.details &&
                            graphQLError.details[fieldName]
                    ) as string[] | undefined);
                if (messages) {
                    return messages[0];
                }
            }
            return false;
        },
        [error]
    );
};
