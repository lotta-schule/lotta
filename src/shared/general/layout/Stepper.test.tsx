import userEvent from '@testing-library/user-event';
import { render } from 'test/util';
import { Stepper } from './Stepper';

describe('Stepper', () => {
    it('should render a stepper', () => {
        const screen = render(
            <Stepper currentStep={0} maxSteps={3} onStep={jest.fn()} />
        );
        expect(screen.container).toMatchSnapshot();
    });

    it('should render the correct step count', () => {
        const screen = render(
            <Stepper currentStep={0} maxSteps={3} onStep={jest.fn()} />
        );
        expect(screen.getByText('1 / 3')).toBeVisible();
    });

    describe('previous step', () => {
        it('should select the previous step on button click', () => {
            const onStep = jest.fn();
            const screen = render(
                <Stepper currentStep={1} maxSteps={3} onStep={onStep} />
            );
            userEvent.click(screen.getByRole('button', { name: /vorherig/i }));
            expect(onStep).toHaveBeenCalledWith(0);
        });

        it('should disable previous button when on first step', () => {
            const screen = render(
                <Stepper currentStep={0} maxSteps={3} onStep={jest.fn()} />
            );
            expect(
                screen.getByRole('button', { name: /vorherig/i })
            ).toBeDisabled();
        });
    });

    describe('next step', () => {
        it('should select the next step on button click', () => {
            const onStep = jest.fn();
            const screen = render(
                <Stepper currentStep={1} maxSteps={3} onStep={onStep} />
            );
            userEvent.click(screen.getByRole('button', { name: /nächst/i }));
            expect(onStep).toHaveBeenCalledWith(2);
        });

        it('should disable previous button when on first step', () => {
            const screen = render(
                <Stepper currentStep={2} maxSteps={3} onStep={jest.fn()} />
            );
            expect(
                screen.getByRole('button', { name: /nächst/i })
            ).toBeDisabled();
        });
    });
});
