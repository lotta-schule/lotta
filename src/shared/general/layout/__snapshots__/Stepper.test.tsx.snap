// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper should render a stepper 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <div
      class="root"
    >
      <button
        aria-label="Vorheriger Schritt"
        aria-selected="false"
        class="lotta-base-button variant__default lotta-button lotta-navigation-button small"
        disabled=""
        role="button"
        type="button"
      >
        <span
          class="lotta-button_icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
            />
          </svg>
        </span>
        <span
          class="lotta-button_text"
        >
          vorheriges
        </span>
      </button>
      <div>
        1
         / 
        3
      </div>
      <button
        aria-label="Nächster Schritt"
        aria-selected="false"
        class="lotta-base-button variant__default lotta-button lotta-navigation-button small"
        role="button"
        type="button"
      >
        <span
          class="lotta-button_icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
            />
          </svg>
        </span>
        <span
          class="lotta-button_text"
        >
          nächstes
        </span>
      </button>
    </div>
  </div>
</div>
`;
